cloud/ec2_price.go

```diff
@@ -93,85 +94,154 @@ func getZone(ctx context.Context, client AWSClient, h *host.Host) (string, error
+				Field: aws.String("instanceType"),
+				Type:  aws.String(match),
+				Value: aws.String(info.instance),
+			},
+			{
+				Field: aws.String("operatingSystem"),
+				Type:  aws.String(match),
+				Value: aws.String(info.os),
+			},
+			{
+				Field: aws.String("location"),
+				Type:  aws.String(match),
+				Value: aws.String(info.region),
+			},
+		},
+		ServiceCode: aws.String("AmazonEC2"),
+	}
+	return constructGetProductsInput
+}
 
-		if err = json.NewDecoder(resp.Body).Decode(&details); err != nil {
-			return true, errors.Wrap(err, "parsing response body")
+// Parse a JSON document like this one
+// {
+// 	"version": "20181031070014",
+// 	"terms": {
+// 		"OnDemand": {
+// 			"XCH2WN4F4MYH63N8.JRTCKXETXF": {
+// 				"termAttributes": {
+// 				},
+// 				"sku": "XCH2WN4F4MYH63N8",
+// 				"priceDimensions": {
+// 					"XCH2WN4F4MYH63N8.JRTCKXETXF.6YS6EN2CT7": {
+// 						"unit": "Hrs",
+// 						"rateCode": "XCH2WN4F4MYH63N8.JRTCKXETXF.6YS6EN2CT7",
+// 						"pricePerUnit": {
+// 							"USD": "0.8400000000"
+// 						},
+// 						"endRange": "Inf",
+// 						"description": "$0.840 per Unused Reservation Linux c3.4xlarge Instance Hour",
+// 						"beginRange": "0",
+// 						"appliesTo": []
+// 					}
+// 				},
+// 				"offerTermCode": "JRTCKXETXF",
+// 				"effectiveDate": "2018-10-01T00:00:00Z"
+// 			}
+// 		}
+// 	},
+// [...]
+func (cpf *cachingPriceFetcher) parseAWSPricing(out *pricing.GetProductsOutput) (float64, error) {
+	if len(out.PriceList) != 1 {
+		return 0, errors.Errorf("problem parsing price list %v", out.PriceList)
+	}
+	terms, ok := out.PriceList[0]["terms"].(map[string]interface{})
+	if !ok {
+		return 0, errors.Errorf("problem parsing price list %v", out.PriceList)
+	}
+	onDemand, ok := terms["OnDemand"].(map[string]interface{})
+	if !ok {
+		return 0, errors.Errorf("problem parsing price list %v", out.PriceList)
+	}
+	var priceDimensions map[string]interface{}
+	for _, v := range onDemand {
+		priceDimensions, ok = v.(map[string]interface{})["priceDimensions"].(map[string]interface{})
+		if !ok {
+			return 0, errors.Errorf("problem parsing price list %v", out.PriceList)
 		}
-		return false, nil
-	}, awsClientImplRetries, awsClientImplStartPeriod)
-
-	if err != nil {
-		return errors.WithStack(err)
 	}
-
-	for _, p := range details.Products {
-		if p.ProductFamily == "Compute Instance" &&
-			p.Attributes.PreInstalledSW == "NA" &&
-			p.Attributes.Tenancy == "Shared" &&
-			p.Attributes.LicenseModel != "Bring your own license" {
-			// the product description does not include pricing information,
-			// so we must look up the SKU in the "Terms" section.
-			price := details.Terms.skuPrice(p.SKU)
-			cpf.ec2Prices[odInfo{
-				os:       p.Attributes.OperatingSystem,
-				instance: p.Attributes.InstanceType,
-				region:   p.Attributes.Location,
-			}] = price
+	var price string
+	for _, v := range priceDimensions {
+		price, ok = v.(map[string]interface{})["pricePerUnit"].(map[string]interface{})["USD"].(string)
```

Comment:
probably need to break this into 4 statements.

Reply:
done